core/storage/couchStorageHelpers.go

```diff
@@ -0,0 +1,582 @@
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+	row := db.Get(context.TODO(), id)
+	if kivik.StatusCode(row.Err) == http.StatusNotFound {
+		return 0, notFound
+	}
+
+	content := ioutil.NopCloser(dataReader)
+	defer content.Close()
+
+	attachment := &kivik.Attachment{Filename: id, ContentType: "application/octet-stream", Content: content}
+	_, err := db.PutAttachment(context.TODO(), id, row.Rev, attachment)
+	if err != nil {
+		return 0, err
+	}
+
+	attachmentMeta, err := db.GetAttachmentMeta(context.TODO(), id, id)
+	if err != nil {
+		return 0, err
+	}
+
+	return attachmentMeta.Size, nil
+}
+
+func (store *CouchStorage) removeAttachment(id string) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+	row := db.Get(context.TODO(), id)
+	if kivik.StatusCode(row.Err) == http.StatusNotFound {
+		return notFound
+	}
+
+	if _, err := db.DeleteAttachment(context.TODO(), id, row.Rev, id); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) upsertObject(id string, object interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	if _, err := db.Put(context.TODO(), id, object); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) getInstanceID() int64 {
+	return time.Now().UnixNano() / (int64(time.Millisecond) / int64(time.Nanosecond))
+}
+
+func (store *CouchStorage) findAll(query interface{}, result interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return notFound
+		}
+		return err
+	}
+
+	resultv := reflect.ValueOf(result)
+	if resultv.Kind() != reflect.Ptr {
+		return &Error{"result argument must be a slice address"}
+	}
+
+	slicev := resultv.Elem()
+
+	if slicev.Kind() == reflect.Interface {
+		slicev = slicev.Elem()
+	}
+	if slicev.Kind() != reflect.Slice {
+		return &Error{"result argument must be a slice address"}
+	}
+
+	slicev = slicev.Slice(0, slicev.Cap())
+	elemt := slicev.Type().Elem()
+	i := 0
+	for rows.Next() {
+		elemp := reflect.New(elemt)
+		err := rows.ScanDoc(elemp.Interface())
+		if err != nil {
+			return err
+		}
+
+		slicev = reflect.Append(slicev, elemp.Elem())
+		slicev = slicev.Slice(0, slicev.Cap())
+		i++
+	}
+	resultv.Elem().Set(slicev.Slice(0, i))
+
+	if err := rows.Err(); err != nil {
+		return err
+	}
+	return rows.Close()
+}
+
+func (store *CouchStorage) countObjects(query interface{}) (uint32, common.SyncServiceError) {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return 0, nil
+		}
+		return 0, err
+	}
+
+	var count uint32
+	for rows.Next() {
```

Comment:
Sometimes databases have built in ability to count a result set and just return you the number of rows. It's more efficient than returning all the rows an counting them. Postgresql, for example, has the COUNT verb that can be used on a query. Does couch have this ability?

Reply:
I did research about this a lot, but I could not find an in-built way to count objects that satisfy a query in couch. Any other implementation I found involved creating a view, which is what _find does as well...

Reply:
ok